client/network/src/request_responses.rs

```diff
@@ -345,148 +351,209 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// One of our requests has failed.
-						RequestResponseEvent::OutboundFailure {
+					// One of our requests has failed.
+					RequestResponseEvent::OutboundFailure {
+						request_id,
+						error,
+						..
+					} => {
+						let out = Event::RequestFinished {
 							request_id,
-							error,
-							..
-						} => {
-							let out = Event::RequestFinished {
-								request_id,
-								result: Err(RequestFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
-						}
+							result: Err(RequestFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// Remote has tried to send a request but failed.
-						RequestResponseEvent::InboundFailure { peer, error, .. } => {
-							let out = Event::InboundRequest {
-								peer,
-								protocol: protocol.clone(),
-								result: Err(ResponseFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					// Remote has tried to send a request but failed.
+					RequestResponseEvent::InboundFailure { request_id, peer, error, .. } => {
+						if self.pending_responses_arrival_time.remove(&request_id).is_none() {
+							debug_assert!(
+								false,
+								"Expected to find start time for failed response.",
+							)
+						}
+						let out = Event::InboundRequest {
+							peer,
+							protocol: protocol.clone(),
+							result: Err(ResponseFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
+					RequestResponseEvent::ResponseSent { request_id, peer } => {
+						match self.pending_responses_arrival_time.remove(&request_id) {
+							Some(arrival) => {
+								let out = Event::InboundRequest {
+									peer,
+									protocol: protocol.clone(),
+									result: Ok(arrival.elapsed()),
+								};
+								return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+							},
+							None => debug_assert!(
+								false,
+								"Expected to find start time for sent response.",
+							),
 						}
-					};
-				}
-			}
 
-			break Poll::Pending;
+					}
+				};
+			}
 		}
+
+		// Sending a response can fail silently if the connection to the peer closed after calling
+		// [`RequestResponse::send_response`] but before the response could be sent on the
+		// connection. Clean up arrival times for responses that have silently failed.
+		let now = Instant::now();
+		self.pending_responses_arrival_time.retain(|request_id, arrival_time| {
+			if now.duration_since(*arrival_time) > Duration::from_secs(10 * 60) {
+				log::warn!(
+					target: "sub-libp2p",
+					"Connection to peer closed after calling `send_response` but before response \
+					 could be send on the connection. Response {:?} has been dropped.",
+					request_id,
+				);
```

Comment:
One could as well emit an event here, in order for these silently failing requests to be tracked via Prometheus as well. I don't think this case is going to be hit very often, thus not worth the additional data to be tracked (request id and protocol). In case it does and we see the warning in the wild, we can still emit the event later on.